astropy/units/structured.py

```diff
@@ -0,0 +1,410 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+This module defines structured units and quantities.
+"""
+
+# Standard library
+import operator
+
+import numpy as np
+
+from .core import Unit, UnitBase
+
+
+__all__ = ['StructuredUnit']
+
+
+def _names_from_dtype(dtype):
```

Comment:
Why doesn't ``https://numpy.org/devdocs/reference/generated/numpy.dtype.names.html`` work?

Reply:
That just gives the names of the upper level - this one is recursive. But I've rewritten it to use `dtype.names` which should make it a bit clearer.